File: src/common/ConsistentHash.cpp
#include "ConsistentHash.h"
#include <iostream>

using namespace std;

size_t ConsistentHash::HashKey(const string &key) const
{
    // Using std::hash for simplicity
    // In production, you might want a better distributed hash like MurmurHash
    hash<string> hasher;
    return hasher(key);
}

void ConsistentHash::AddNode(const string &nodeAddress)
{
    size_t pos = HashKey(nodeAddress);
    ring[pos] = nodeAddress;
}

void ConsistentHash::RemoveNode(const string &nodeAddress)
{
    size_t pos = HashKey(nodeAddress);
    ring.erase(pos);
}

string ConsistentHash::GetNodeForKey(const string &key) const
{
    if (ring.empty())
    {
        return "";
    }

    size_t pos = HashKey(key);

    // Find the first node with position >= key's position
    // This is "walking clockwise" on the ring
    auto it = ring.lower_bound(pos);

    // If we've gone past the end, wrap around to the first node
    if (it == ring.end())
    {
        it = ring.begin();
    }

    return it->second;
}

vector<string> ConsistentHash::GetNodesForKey(const string &key, int replicaCount) const
{
    vector<string> nodes;

    if (ring.empty() || replicaCount <= 0)
    {
        return nodes;
    }

    // Don't ask for more replicas than we have nodes
    int maxReplicas = min(replicaCount, (int)ring.size());

    size_t pos = HashKey(key);

    // Find the first node (starting point)
    auto it = ring.lower_bound(pos);
    if (it == ring.end())
    {
        it = ring.begin();
    }

    // Walk clockwise and collect nodes
    while ((int)nodes.size() < maxReplicas)
    {
        nodes.push_back(it->second);

        // Move to next node
        ++it;

        // Wrap around if we hit the end
        if (it == ring.end())
        {
            it = ring.begin();
        }
    }

    return nodes;
}

vector<string> ConsistentHash::GetAllNodes() const
{
    vector<string> nodes;
    for (const auto &pair : ring)
    {
        nodes.push_back(pair.second);
    }
    return nodes;
}

size_t ConsistentHash::GetNodeCount() const
{
    return ring.size();
}

bool ConsistentHash::HasNode(const string &nodeAddress) const
{
    size_t pos = HashKey(nodeAddress);
    return ring.find(pos) != ring.end();
}

void ConsistentHash::PrintRing() const
{
    cout << "=== Consistent Hash Ring ===" << endl;
    cout << "Nodes: " << ring.size() << endl;

    for (const auto &pair : ring)
    {
        cout << "  Position " << pair.first << " → " << pair.second << endl;
    }

    cout << "=============================" << endl;
}

File: src/main/java/com/distributed/storage/ConsistentHash.java
package com.distributed.storage;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class ConsistentHash {
    // The ring: maps position (hash value) -> node address
    private TreeMap<Integer, String> ring = new TreeMap<>();

    // Hash function to map strings to positions on the ring
    private int hashKey(String key) {
        // Using String.hashCode for simplicity
        // In production, you might want a better distributed hash like MurmurHash
        return key.hashCode();
    }

    // Add a node to the ring
    public void addNode(String nodeAddress) {
        int pos = hashKey(nodeAddress);
        ring.put(pos, nodeAddress);
    }

    // Remove a node from the ring
    public void removeNode(String nodeAddress) {
        int pos = hashKey(nodeAddress);
        ring.remove(pos);
    }

    // Get the node responsible for a given key (CID)
    public String getNodeForKey(String key) {
        if (ring.isEmpty()) {
            return "";
        }

        int pos = hashKey(key);

        // Find the first node with position >= key's position
        // This is "walking clockwise" on the ring
        Map.Entry<Integer, String> entry = ring.ceilingEntry(pos);

        // If we've gone past the end, wrap around to the first node
        if (entry == null) {
            entry = ring.firstEntry();
        }

        return entry.getValue();
    }

    // Get all nodes in the ring
    public List<String> getAllNodes() {
        return new ArrayList<>(ring.values());
    }

    // Get number of nodes
    public int getNodeCount() {
        return ring.size();
    }

    // Check if a node exists in the ring
    public boolean hasNode(String nodeAddress) {
        int pos = hashKey(nodeAddress);
        return ring.containsKey(pos);
    }

    // Debug: print the ring structure
    public void printRing() {
        System.out.println("=== Consistent Hash Ring ===");
        System.out.println("Nodes: " + ring.size());

        for (Map.Entry<Integer, String> entry : ring.entrySet()) {
            System.out.println("  Position " + entry.getKey() + " -> " + entry.getValue());
        }

        System.out.println("=============================");
    }
}
